File: Context/GameSpec.StoreManager/StoreManagers/BlizzardStoreManager.cs
using Google.Protobuf;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text.Json;

namespace GameSpec.StoreManagers
{
    /// <summary>
    /// BlizzardStoreManager
    /// </summary>
    internal static class BlizzardStoreManager
    {
        static Dictionary<string, string> AppPaths = new Dictionary<string, string>();

        public static bool TryGetPathByKey(string key, JsonProperty prop, JsonElement? keyElem, out string path)
            => AppPaths.TryGetValue(key, out path);

        static BlizzardStoreManager()
        {
            var root = GetPath();
            if (root == null) return;
            var dbPath = Path.Combine(root, "product.db");
            if (!File.Exists(dbPath)) return;
            using var s = File.OpenRead(dbPath);
            Database data;
            try
            {
                data = Database.Parser.ParseFrom(s);
            }
            catch (InvalidProtocolBufferException)
            {
                data = new Database { ProductInstall = { ProductInstall.Parser.ParseFrom(s) } };
            }
            foreach (var app in data.ProductInstall)
            {
                var appPath = app.Settings.InstallPath;
                if (Directory.Exists(appPath)) AppPaths.Add(app.Uid, appPath);
            }
        }

        static string GetPath()
        {
            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), "Battle.net", "Agent");
            else if (RuntimeInformation.OSDescription.StartsWith("android-")) return null;
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
            {
                var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
                return new[] { ".steam", ".steam/steam", ".steam/root", ".local/share/Steam" }
                    .Select(path => Path.Join(home, path, "appcache"))
                    .FirstOrDefault(Directory.Exists);
            }
            else if (RuntimeInformation.IsOSPlatform(OSPlatform.OSX))
            {
                var home = "/Users/Shared";
                return new[] { "Battle.net/Agent" }
                    .Select(path => Path.Join(home, path, "data"))
                    .FirstOrDefault(Directory.Exists);
            }
            throw new PlatformNotSupportedException();
        }
    }
}

File: Applications/Book/book/03-locate-files/python/BlizzardStoreManager.py
import os, platform, BlizzardProtoDatabase_pb2

def init():
    def getPath():
        system = platform.system()
        if system == 'Windows':
            return os.path.join(os.getenv('ALLUSERSPROFILE'), 'Battle.net', 'Agent')
        elif system == 'Linux':
            home = os.path.expanduser('~')
            paths = [os.path.join(home, path, 'appcache') for path in ['.steam', '.steam/steam', '.steam/root', '.local/share/Steam']]
            return next(iter(x for x in paths if os.path.isdir(x)), None)
        elif system == 'Darwin':
            home = '/Users/Shared'
            paths = [os.path.join(home, path, 'data') for path in ['Battle.net/Agent']]
            return next(iter(x for x in paths if os.path.isdir(x)), None)
        else: raise Exception(f'Unknown platform: {system}')
    root = getPath()
    if root == None: return
    dbPath = os.path.join(root, 'product.db')
    if not os.path.exists(dbPath): return
    productDb = BlizzardProtoDatabase_pb2.Database()
    with open(dbPath, 'rb') as f:
        bytes = f.read()
        productDb.ParseFromString(bytes)
        #try: database.ParseFromString(bytes)
        #except InvalidProtocolBufferException: return None
        for app in productDb.ProductInstall:
            appPath = app.Settings.InstallPath
            if os.path.isdir(appPath): blizzardAppPaths[app.Uid] = appPath

blizzardAppPaths = {}
init()
print(blizzardAppPaths)
